docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,223 @@
+    - `rescheduleCriticalSeconds`: the critical duration that controller will trigger a Pod to reschedule to another subset.
+      > rescheduleCriticalSeconds indicates how long controller will reschedule a schedule failed Pod to the subset that has
+        redundant capacity after the subset where the Pod lives. \
+        If a Pod was scheduled failed and still in a pending status over RescheduleCriticalSeconds duration, the controller will reschedule it to the suitable subset.
+
+**status**:
+- `subsetStatuses`: statuses of subsets defined in spec
+  - `missingReplicas`: the missing number of this subset, most of the time it is (maxReplicas - the number of active Pods)
+    > missingReplicas > 0 indicates the subset is still missing MissingReplicas pods to create \
+    missingReplicas = 0 indicates the subset already has enough pods, there is no need to create \
+    missingReplicas = -1 indicates the subset's MaxReplicas not set, then there is no limit for pods number
+  - `creatingPods`: the map of Pods that just allocated to this subset by webhook
+    > creatingPods contains information about pods whose creation was processed by
+      the webhook handler but not yet been observed by the WorkloadSpread controller.
+      A pod will be in this map from the time when the webhook handler processed the
+      creation request to the time when the pod is seen by controller. \
+      The key in the map is the name of the pod and the value is the time when the webhook
+      handler process the creation request. If the real creation didn't happen and a pod is
+      still in this map, it will be removed from the list automatically by WorkloadSpread controller
+      after some time.\
+      If everything goes smooth this map should be empty for the most of the time.
+      Large number of entries in the map may indicate problems with pod creations.
+  - `deletingPods`: the map of Pods been deleted or evicted that just noticed by webhook
+    > deletingPods is similar with creatingPods and it contains information about pod deletion.
+  - `subsetUnscheduledStatus`: contains the details for the unscheduled subset.
+     - `unschedulable`: unschedulable is true indicates this Subset cannot be scheduled. The default is false.
+     - `unscheduledTime`: last time for unscheduled.
+     - `failedCount`: the number of subset was marked with unschedulable.
+
+### Requirements
+
+WorkloadSpread requires Pod webhook to inject rules into Pods and notice the deletion and eviction of Pods.
+
+So if the `PodWebhook` feature-gate is set to `false`, WorkloadSpread will also be disabled.
+
+### Implementation Details/Notes/Constraints
+
+WorkloadSpread have both webhook and controller. Controller should collaborate with webhook to maintain WorkloadSpread status together. \
+The controller is responsible for calculating the real status, and the webhook mainly counts missingReplicas and records the creation or deletion entry of Pod into map.
```

Comment:
The webhook is responsible for injecting rules into pod, reducing missingReplicas and recording the creation or deletion entry of Pod into map.
The controller is responsible for updating the missingReplicas along with other statics, and clean creatingpods and deletingPods.